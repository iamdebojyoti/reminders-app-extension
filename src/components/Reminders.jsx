import { Component } from "react";
import { Tab, Tabs } from "react-bootstrap";
import { v4 as uuidv4 } from 'uuid';
import Input from "./Input";
import Upcoming from "./Upcoming";
import Done from "./Done";

class Reminders extends Component {

    constructor(props) {
        super(props);
        this.state = {
            remindersMap: new Map(),
            reminders: []
        }
        this.addReminder = this.addReminder.bind(this);
        this.complete = this.complete.bind(this);
        this.remove = this.remove.bind(this);
        this.upcoming = this.upcoming.bind(this);
        this.done = this.done.bind(this);
    }
    
    addReminder(reminder) {
        reminder.id = uuidv4();
        reminder.state = "CREATED";
        
        let currentReminders = this.state.remindersMap;
        currentReminders.set(reminder.id, reminder);
        this.setState({reminders: [...currentReminders.values()]});
    }

    complete(id) {
        let currentReminders = this.state.remindersMap;
        let reminder = currentReminders.get(id);
        reminder.state = "DONE";
        currentReminders.set(reminder.id, reminder);
        this.setState({reminders: [...currentReminders.values()]});
    }

    remove(id) {
        let currentReminders = this.state.remindersMap;
        currentReminders.delete(id);
        this.setState({reminders: [...currentReminders.values()]});
    }

    upcoming() {
        return this.state.reminders.filter(reminder => reminder.state === 'CREATED')
    }

    done() {
        return this.state.reminders.filter(reminder => reminder.state === 'DONE')
    }

    render() {
        return (
            <Tabs defaultActiveKey="input" id="uncontrolled-tab-example" className="mb-3">
                <Tab eventKey="input" title="Input">
                    <Input addReminder={this.addReminder}/>
                </Tab>
                <Tab eventKey="upcoming" title="Upcoming">
                    <Upcoming reminders={this.upcoming()} complete={this.complete}/>
                </Tab>
                <Tab eventKey="done" title="Done">
                    <Done reminders={this.done()} remove={this.remove}/>
                </Tab>
            </Tabs>
        );
    }
}

export default Reminders;