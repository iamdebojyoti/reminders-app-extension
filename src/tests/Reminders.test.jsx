import { shallow } from "enzyme";
import Reminders from "../components/Reminders";

describe('Testing <Reminders/> component', ()=> {

    let component;
    it('should match snapshot of Reminders', ()=> {
      component = shallow(<Reminders/>);
      expect(component.debug()).toMatchSnapshot();
    });

    it('should be able to add reminders', () => {
      component = shallow(<Reminders/>);
      component.instance().addReminder({name:"Name1", desc: "Desc1"});
      
      const state = component.state('remindersMap');
      const [reminder] = state.values();
      
      expect(reminder.name).toBe("Name1");
      expect(reminder.desc).toBe("Desc1");
      expect(reminder.state).toBe("CREATED");
    });

    it('should be able to add multiple reminders', () => {
      component = shallow(<Reminders/>);
      component.instance().addReminder({name:"Name1", desc: "Desc1"});
      component.instance().addReminder({name:"Name2", desc: "Desc2"});
      
      const state = component.state('remindersMap');
      const [reminder1, reminder2] = state.values();
      
      expect(reminder1.name).toBe("Name1");
      expect(reminder1.desc).toBe("Desc1");
      expect(reminder1.state).toBe("CREATED");

      expect(reminder2.name).toBe("Name2");
      expect(reminder2.desc).toBe("Desc2");
      expect(reminder2.state).toBe("CREATED");
    });

    it('should be able to complete a reminder', () => {
      component = shallow(<Reminders/>);
      component.instance().addReminder({name:"Name1", desc: "Desc1"});
      component.instance().addReminder({name:"Name2", desc: "Desc2"});
      
      const state = component.state('remindersMap');
      const [reminder1, reminder2] = state.values();
      
      component.instance().complete(reminder1.id);
      expect(reminder1.name).toBe("Name1");
      expect(reminder1.desc).toBe("Desc1");
      expect(reminder1.state).toBe("DONE");

      expect(reminder2.name).toBe("Name2");
      expect(reminder2.desc).toBe("Desc2");
      expect(reminder2.state).toBe("CREATED");
    });

    it('should be able to remove a reminder', () => {
      component = shallow(<Reminders/>);
      component.instance().addReminder({name:"Name1", desc: "Desc1"});
      component.instance().addReminder({name:"Name2", desc: "Desc2"});
      
      const state = component.state('remindersMap');
      const [reminder1] = state.values();
      
      component.instance().remove(reminder1.id);
      expect(state.size).toBe(1);
    });

  });