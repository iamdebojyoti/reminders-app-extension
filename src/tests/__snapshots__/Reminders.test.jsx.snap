// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Reminders/> component should match snapshot of Reminders 1`] = `
"<Tabs defaultActiveKey=\\"input\\" id=\\"uncontrolled-tab-example\\" className=\\"mb-3\\" variant=\\"tabs\\" mountOnEnter={false} unmountOnExit={false}>
  <Tab eventKey=\\"input\\" title=\\"Input\\">
    <Input addReminder={[Function: bound addReminder]} />
  </Tab>
  <Tab eventKey=\\"upcoming\\" title=\\"Upcoming\\">
    <Upcoming reminders={{...}} complete={[Function: bound complete]} />
  </Tab>
  <Tab eventKey=\\"done\\" title=\\"Done\\">
    <Done reminders={{...}} remove={[Function: bound remove]} />
  </Tab>
</Tabs>"
`;
